### PR TITLE
planner: fix panic while get part of partition key values

### LINKED ISSUE #22396: tidb panic while get part of partition key values
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
```sql
CREATE TABLE tbl_311 ( COL1 INT, COL2 VARCHAR(20), COL3 bigint, COL4 FLOAT, COL5 DATETIME, primary key (COL1, col2, col3)) PARTITION BY RANGE (COL1 + COL3) (
    PARTITION p0 VALUES LESS THAN (0),
    PARTITION p1 VALUES LESS THAN(10),
    PARTITION p2 VALUES LESS THAN(20),
    PARTITION p3 VALUES LESS THAN MAXVALUE
);
insert into tbl_311(col1, col2, col3) values (1, 'a', 11), (2, 'b', 22), (3, 'c', 33), (10, 'd', 44), (9, 'e', 55);
select * from tbl_311 where col1 = 1;

```
### 2. What did you expect to see? (Required)
```sql
mysql > select * from tbl_311 where col1 = 1;
+------+------+------+--------+--------+
| COL1 | COL2 | COL3 | COL4   | COL5   |
+------+------+------+--------+--------+
| 1    | a    | 11   | <null> | <null> |
+------+------+------+--------+--------+
```
### 3. What did you see instead (Required)
tidb panic
```sql
mysql > select * from tbl_311 where col1 = 1
(1105, 'runtime error: invalid memory address or nil pointer dereference')
```
### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
```
mysql root@127.0.0.1:pkq> select tidb_version();
Reconnecting...
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                    |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-2016-g3dd842f50\nEdition: Community\nGit Commit Hash: 3dd842f50a75ebbf9a4f9d7b30fb2ce8a8dd4b37\nGit Branch: master\nUTC Build Time: 2021-01-14 03:27:59\nGoVersion: go1.15\nRace Enabled: false\nTiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306\nCheck Table Before Drop: false |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

1 row in set
Time: 0.010s

```

### PR DESCRIPTION
**Pull request by beihaiguaishou**
### What problem does this PR solve?

Issue Number: close #22396 

Problem Summary: TiDB will panic while getting part of partition key values, such as we partition by the expression "a + b", it will panic while we run a query with condition that reference column a or b.

### What is changed and how it works?

What's Changed: Before evolute the value of the partition expression, We check whether the expression can evolute.

How it Works: This issue occur after this pr [https://github.com/pingcap/tidb/pull/19267](url). Before this pr, the code in `makePartitionByFnCol`:
```go
if _, ok := monotoneIncFuncs[raw.FuncName.L]; ok {
   fn = raw
   args := fn.GetArgs()
   if len(args) > 0 {
      arg0 := args[0]
      if c, ok1 := arg0.(*expression.Column); ok1 {
         col = c
      }
   }
}
```
After the pr, the code is:
```go
fn = raw
args := fn.GetArgs()
if len(args) > 0 {
   arg0 := args[0]
   if c, ok1 := arg0.(*expression.Column); ok1 {
      col = c
   }
}
```
It removes the condition and causes makePartitionByFnCol return col even if this function takes more than one variable input. And finally, the code runs to `extractDataForPrune`, and the code `constExpr.EvalInt(sctx, chunk.Row{})` panic because of lack of inputs.
And I think we can do more while the partition expression is related to more than one column. This pr just add a check before evolute the value of the partition expression.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test

Side effects

- None

### Release note 

- No release note

### CODE REVIEW

File: planner/core/rule_partition_processor.go [hunk 1]
```diff
@@ -978,6 +978,9 @@ func (p *rangePruner) extractDataForPrune(sctx sessionctx.Context, expr expressi
 		// If the partition expression is col, use constExpr.
 		constExpr = con
 	}
+	if !constExpr.ConstItem(sctx.GetSessionVars().StmtCtx) {
```
Comment (eurekaka) [inline planner/core/rule_partition_processor.go:981]: Please add a comment here to include the cases in the issue and inform that `constExpr` may not really be a constant when coming here.

**Review by eurekaka — COMMENTED** [review]
LGTM

File: planner/core/rule_partition_processor.go (hunk 1 above)
Comment (beihaiguaishou) [reply planner/core/rule_partition_processor.go:985]: added

**Review by crazycs520 — COMMENTED** [review]
/LGTM

**Pull request merged**